pkg/cri/resource-manager/policy/builtin/podpools/metrics.go

```diff
@@ -0,0 +1,242 @@
+	if err != nil {
+		return nil, err
+	}
+	return append(threadCPUMetrics, poolCPUMetrics...), nil
+}
+
+// updateThreadCPUUsageMetrics collects the CPU usage of each thread.
+func updateThreadCPUUsageMetrics() []prometheus.Metric {
+	cpuTimeStat.RLock()
+	defer cpuTimeStat.RUnlock()
+	threadCPUUsage := make([]prometheus.Metric, cpuNumber)
+	for i := 0; i < cpuNumber; i++ {
+		threadCPUUsage[i] = prometheus.MustNewConstMetric(
+			descriptors[threadCPUUsageDesc],
+			prometheus.GaugeValue,
+			cpuTimeStat.CPUUsage[i],
+			strconv.Itoa(i),
+		)
+	}
+	return threadCPUUsage
+}
+
+// updatePoolCPUUsageMetrics collects the CPU usage of pools defined by podpools-policy.
+func updatePoolCPUUsageMetrics(ppm *Metrics) ([]prometheus.Metric, error) {
+	if ppm == nil {
+		return nil, fmt.Errorf("Podpools metrics used to count pool CPU usage is missing.")
+	}
+	// Sort the pool metrics.
+	poolNames := make([]string, 0, len(ppm.PoolMetrics))
+	for poolName := range ppm.PoolMetrics {
+		poolNames = append(poolNames, poolName)
+	}
+	sort.Sort(sort.StringSlice(poolNames))
+
+	// Calculate the CPU usage of a pool and send to prometheus.
+	poolCPUUsageMetrics := make([]prometheus.Metric, len(poolNames))
+	poolCPUUsageList := make(map[string]float64, len(poolNames))
+	cpuTimeStat.RLock()
+	defer cpuTimeStat.RUnlock()
+	for index, poolName := range poolNames {
+		cpus := resolvePools(ppm.PoolMetrics[poolName].CPUs)
+		cpusPerPoolList := strings.Split(cpus, ",")
+		poolDeltaIdleTime := uint64(0)
+		poolDeltaTotalTime := uint64(0)
+		for _, cpuId := range cpusPerPoolList {
+			cpuIdInt, _ := strconv.Atoi(cpuId)
+			poolDeltaIdleTime += cpuTimeStat.DeltaIdleTime[cpuIdInt]
+			poolDeltaTotalTime += cpuTimeStat.DeltaTotalTime[cpuIdInt]
+		}
+		poolCPUUsageList[poolName] = (1.0 - float64(poolDeltaIdleTime)/float64(poolDeltaTotalTime)) * 100.0 * float64(len(cpusPerPoolList))
+		poolCPUUsageMetrics[index] = prometheus.MustNewConstMetric(
+			descriptors[poolCPUUsageDesc],
+			prometheus.GaugeValue,
+			poolCPUUsageList[poolName],
+			PolicyName,
+			poolName,
+			ppm.PoolMetrics[poolName].DefName,
+			ppm.PoolMetrics[poolName].CPUs,
+			ppm.PoolMetrics[poolName].Memory,
+			ppm.PoolMetrics[poolName].CPUMiliSize,
+			ppm.PoolMetrics[poolName].PodNames,
+			ppm.PoolMetrics[poolName].ContainerNames,
+		)
+	}
+	return poolCPUUsageMetrics, nil
+}
+
+// resolvePools resolves pools' cpuset into single cpus.
+func resolvePools(cpuset string) string {
+	cpusetArray := strings.Split(cpuset, ",")
+	cpus := ""
+	for _, cpusetStr := range cpusetArray {
+		cpuMember := strings.Split(cpusetStr, "-")
+		if cpus != "" {
+			cpus = fmt.Sprintf(cpus+",%s", cpuMember[0])
+		} else {
+			cpus = fmt.Sprintf("%s", cpuMember[0])
+		}
+		if len(cpuMember) > 1 {
+			begin, _ := strconv.Atoi(cpuMember[0])
+			end, _ := strconv.Atoi(cpuMember[1])
+			for j := begin + 1; j <= end; j++ {
+				cpus = fmt.Sprintf(cpus+",%d", j)
+			}
+		}
+	}
+	return cpus
```

Comment:
Since you seem to eventually end up converting the calculated result here to a slice of ints, it would be much simpler to use the k8s `cpuset` package for that, and directly loop through the resulting slice at the current call site to `resolvePools`.
```
import (
    ...
    "github.com/pkg/errors"
    "k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
    ...
)
...
// It is almost a one-liner...
func splitToCPUs(cpus string) (string, error) {
    if cset, err := cpuset.Parse(cpus); err != nil {
        return "", errors.Wrapf(err, "invalid cpuset %q", cpus)
    } else {
        return cset.ToSlice(), nil
    }
}

// And if you are 100% sure that the passed in cpus(et) string is always a valid cpuset
// string (cpuset.MustParse() panics on errors) then you can really use a one-liner at the
// current call site like this, instead of writing/calling a function:
....
    for _, cpuID := range cpuset.MustParse(cpus).ToSlice() {
        do_the_deed_for_cpuID()
    }
...
```